pkg/services/ngalert/notifier/alertmanager.go

```diff
@@ -32,70 +32,141 @@ type Alertmanager struct {
+	// TODO: check if config is same as before using hashes and skip reload in case they are same.
+	cfg1, cfg2, err := getConfigFromDatabase()
+	if err != nil {
+		return errors.Wrap(err, "get config from database")
+	}
+	return errors.Wrap(am.reloadFromConfig(cfg1, cfg2), "reload from config")
+}
+
+func getConfigFromDatabase() (interface{}, interface{}, error) {
+	// TODO: get configs from the database.
+	return nil, nil, nil
+}
+
+// reloadFromConfig re-initialises all components using the config provided and restarts dispatcher.
+// TODO: replace these dummy configs with actual ones and use them.
+func (am *Alertmanager) reloadFromConfig(cfg1 interface{}, cfg2 interface{}) error {
 	//TODO: Speak with David Parrot wrt to the marker, we'll probably need our own.
-	am.marker = types.NewMarker(prometheus.DefaultRegisterer)
+	marker := types.NewMarker(prometheus.DefaultRegisterer)
 
-	var err error
-	am.silences, err = silence.New(silence.Options{
+	var (
+		err      error
+		silences *silence.Silences
+		alerts   *AlertProvider
+	)
+	silences, err = silence.New(silence.Options{
 		SnapshotFile: filepath.Join("dir", "silences"), //TODO: This is a setting
 		Retention:    time.Hour * 24,                   //TODO: This is also a setting
 	})
 	if err != nil {
 		return errors.Wrap(err, "unable to initialize the silencing component of alerting")
 	}
 
-	am.notificationLog, err = nflog.New(
-		nflog.WithRetention(time.Hour*24),                         //TODO: This is a setting.
-		nflog.WithSnapshot(filepath.Join("dir", "notifications")), //TODO: This should be a setting
-	)
-	if err != nil {
-		return errors.Wrap(err, "unable to initialize the notification log component of alerting")
-	}
-
-	{
-		// Now, let's put together our notification pipeline
-		receivers := buildIntegrationsMap()
-		routingStage := make(notify.RoutingStage, len(receivers))
+	// Now, let's put together our notification pipeline
+	receivers := buildIntegrationsMap()
+	routingStage := make(notify.RoutingStage, len(receivers))
 
-		silencingStage := notify.NewMuteStage(silence.NewSilencer(am.silences, am.marker, gokit_log.NewNopLogger()))
-		//TODO: We need to unify these receivers
-		for name := range receivers {
-			stage := createReceiverStage(name, receivers[name], waitFunc, am.notificationLog)
-			routingStage[name] = notify.MultiStage{silencingStage, stage}
-		}
+	silencingStage := notify.NewMuteStage(silence.NewSilencer(silences, marker, gokit_log.NewNopLogger()))
+	//TODO: We need to unify these receivers
+	for name := range receivers {
+		stage := createReceiverStage(name, receivers[name], waitFunc, am.notificationLog)
+		routingStage[name] = notify.MultiStage{silencingStage, stage}
+	}
 
-		am.alerts, err = NewAlertProvider(routingStage, am.marker, gokit_log.NewNopLogger())
-		if err != nil {
-			return errors.Wrap(err, "failed to initialize alerting storage component")
-		}
+	alerts, err = NewAlertProvider(routingStage, marker, gokit_log.NewNopLogger())
```

Comment:
I believe this one applies as well, move to init as we don't need to recreate it otherwise we'll delete all the alerts.

Reply:
Good point!